### ISSUE TITLE
"No value specified" for non-declared bind parameter

### ISSUE THREAD

**Issue by Queatz** [arangodb/arangodb-java-driver#177]
I'm experiencing a bit of an odd behavior.

The following error is being thrown sporadically (about every 20-50th query):

```java
com.arangodb.ArangoDBException: Response: 400, Error: 1551 - AQL: no value specified for declared bind parameter 'client' (while parsing)
	at com.arangodb.internal.util.ResponseUtils.checkError(ResponseUtils.java:53)
	at com.arangodb.internal.velocystream.VstCommunication.checkError(VstCommunication.java:116)
	at com.arangodb.internal.velocystream.VstCommunicationSync.execute(VstCommunicationSync.java:143)
	at com.arangodb.internal.velocystream.VstCommunicationSync.execute(VstCommunicationSync.java:48)
	at com.arangodb.internal.velocystream.VstCommunication.execute(VstCommunication.java:100)
	at com.arangodb.internal.velocystream.VstProtocol.execute(VstProtocol.java:47)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:71)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:53)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:49)
	at com.arangodb.ArangoDatabase.query(ArangoDatabase.java:358)
	at com.inceptionnotes.sync.store.NoteStore.updateRelationshipsForNoteProp(NoteStore.java:201)
```

The query is:

```java
    private static final String AQL_UPDATE_RELATIONSHIPS_INSERT_STEP = "FOR target IN @value\n" +
            "    UPSERT { _from: @note, _to: target, kind: @prop }\n" +
            "        INSERT { _from: @note, _to: target, kind: @prop }\n" +
            "        UPDATE {}\n" +
            "        IN relationships";
```

As you can observe, there is no `@client` binding in this query.

**Note #1** This is in a multi-threaded websocket environment (running on Tomcat.) 

**Note #2** Other queries using the same `ArangoDatabase` object do use`@client` and it is very likely for those queries to execute close in time to the one throwing the error above.

**Comment by mpv1989** [comment]
Hi @Queatz,

thanks for reporting. This problem belongs to our network protocol VelocyStream. We are currently investigating if the bug is in our Server or the driver.

As long as we have not fixed the error you can use HTTP in driver as a workaround.
```
ArangoDB arango = new ArangoDB.Builder().useProtocol(HTTP_VPACK).build();
```

**Comment by Queatz** [comment]
Thanks @mpv1989 - I went ahead and tested `.useProtocol(Protocol.HTTP_VPACK)` and `.useProtocol(Protocol.HTTP_JSON)` and observed the error with them both.

Here it is over HTTP:
```
16-Feb-2018 23:11:58.399 INFO [https-openssl-nio-0.0.0.0-8443-exec-8] com.inceptionnotes.sync.ws.WebsocketClient.onError WEBSOCKET (ERROR): Response: 400, Error: 1552 - AQL: bind parameter 'prop' was not declared in the query (while parsing)
com.arangodb.ArangoDBException: Response: 400, Error: 1552 - AQL: bind parameter 'prop' was not declared in the query (while parsing)
	at com.arangodb.internal.util.ResponseUtils.checkError(ResponseUtils.java:53)
	at com.arangodb.internal.http.HttpConnection.checkError(HttpConnection.java:340)
	at com.arangodb.internal.http.HttpConnection.execute(HttpConnection.java:204)
	at com.arangodb.internal.http.HttpCommunication.execute(HttpCommunication.java:144)
	at com.arangodb.internal.http.HttpCommunication.execute(HttpCommunication.java:129)
	at com.arangodb.internal.http.HttpProtocol.execute(HttpProtocol.java:49)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:71)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:53)
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:49)
	at com.arangodb.ArangoDatabase.query(ArangoDatabase.java:358)
	at com.inceptionnotes.sync.store.NoteStore.setPropSeenByClient(NoteStore.java:175)
```

**Comment by mpv1989** [comment]
@Queatz Ok, interesting. I wasn't able to reproduce this behavior. Are you able to share parts of your source code which could belong to this problem.

**Comment by Queatz** [comment]
Yes - https://github.com/Queatz/inception-sync/blob/master/sync/src/main/java/com/inceptionnotes/sync/store/NoteStore.java

**Comment by mpv1989** [comment]
@Queatz Thanks.

I think I found the bug. I was able to reproduce your problem if one of the bind parameters has the value `null`. The driver does not serialize null values by default. I changed this for the bind parameters. You can try my fix with driver version `4.3.4-SNAPSHOT`. Please let me know if this fixes your problem.

This fix should also solve your problem https://github.com/arangodb/arangodb/issues/4582

**Comment by Queatz** [comment]
This does fix https://github.com/arangodb/arangodb/issues/4582 I have still seen this error with `4.3.4-SNAPSHOT` When printing out the param bindings, they all are strings.

**Comment by pilot4u** [comment]
I started getting this error, when executing my Foxx service:
AQL: no value specified for declared bind parameter 'lat' (while parsing)

my query:
for rest in Rest
let dist = distance(rest.location.geo[0],rest.location.geo[1], @lat, @lng)
sort dist
return MERGE(rest,{distance:dist})

query parameters:
{"lng":31.749314,"lat":34.9937314}
this issue started today after I upgraded from version 3.3.3 to version 3.3.4

the same query is running OK using arangosh , and arangojs driver.
but when running on Foxx service receiving described error

might it be related to this bug fix, which should be implemented for inner js @arrangodb ?

**Comment by mpv1989** [comment]
Hi @pilot4u ,

as I understand you correctly, in your case the Java driver is not involved. So the problem could be on server side. Can you please open a github issue in https://github.com/arangodb/arangodb with all information about your setup?

Thanks & best
Mark

**Comment by mpv1989** [comment]
released version 4.3.4

**Comment by mpv1989** [comment]
@Queatz if you have still the same problem, please open the issue again and I will take another look
